File: perl/efficiency_nanoseq.pl
#!/usr/bin/env perl

########## LICENCE ##########
# Copyright (c) 2022 Genome Research Ltd
# 
# Author: CASM/Cancer IT <cgphelp@sanger.ac.uk>
# 
# This file is part of NanoSeq.
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
# 
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
# 
# 1. The usage of a range of years within a copyright statement contained within
# this distribution should be interpreted as being equivalent to a list of years
# including the first and last year specified and all consecutive years between
# them. For example, a copyright statement that reads ‘Copyright (c) 2005, 2007-
# 2009, 2011-2012’ should be interpreted as being identical to a statement that
# reads ‘Copyright (c) 2005, 2007, 2008, 2009, 2011, 2012’ and a copyright
# statement that reads ‘Copyright (c) 2005-2012’ should be interpreted as being
# identical to a statement that reads ‘Copyright (c) 2005, 2006, 2007, 2008,
# 2009, 2010, 2011, 2012’.
###########################

use warnings;
use strict;
use Getopt::Long qw(:config no_ignore_case);
use Pod::Usage;
use File::Which;
use Capture::Tiny qw(capture);

# 1. Calculate num reads in the deduplicated bam
# 2. Calculate num reads merged bam
# 3. Calculate duplicate rate and number of bases sequenced
# 4. Get RB conformations - and save
# 5. Calculate efficiency with my own binomial thingy
my %opts = (
    't' => 1,
);

GetOptions('d|dedup=s'   => \my $deduplicated_bam,
           'x|duplex=s'  => \my $merged_bam,
           'o|out=s'     => \my $output_prefix,
           'r|ref=s'     => \my $ref_genome,
           'p|panel=s'   => \my $panel,
           't|threads=i' => \$opts{'t'},
           'h|help'      => \$opts{'h'}
) or pod2usage(2);
pod2usage(-verbose => 1, -exitval => 0) if(defined $opts{'h'});
pod2usage(2) if ( not defined $output_prefix and not defined $deduplicated_bam and not defined $merged_bam and not defined $ref_genome and not defined $panel);

die ("\nOutput prefix not defined\n") unless( $output_prefix );
die ("\nMust define the reference\n") unless ( $ref_genome);
die ("\nReference $ref_genome not found\n") unless ( -e $ref_genome );
die ("\nReference $ref_genome index not found\n") unless ( -e $ref_genome . ".fai" );
die ("\nMust define a deduplicated BAM\n") unless ( $deduplicated_bam );
die ("\nFile $deduplicated_bam not found\n") unless ( -e $deduplicated_bam );
my ($ext) = $deduplicated_bam  =~ /(\.[^.]+)$/;
$ext =~ s/.$/i/;
die ("\nIndex for $deduplicated_bam not found\n") unless ( -e "$deduplicated_bam".$ext );
die ("\nMust define a tumour duplex BAM\n") unless ( $merged_bam );
die ("\nFile $merged_bam not found\n") unless ( -e $merged_bam );
($ext) = $merged_bam  =~ /(\.[^.]+)$/;
$ext =~ s/.$/i/;
die ("\nIndex for $merged_bam $ext not found!\n") unless ( -e "$merged_bam".$ext );
die ("\nsamtools not found in path\n") unless ( which 'samtools' );
die ("\nRscript not found in path\n") unless ( which 'Rscript' );
die ("\nefficiency_nanoseq.R must be in path\n") unless ( which 'efficiency_nanoseq.R' );
my $do_panel;
if ($panel ) {
  die ("\nTargeted panel $panel not found\n") unless( -e $panel );
  $do_panel = 1;
} else {
  $do_panel = 0;
  $panel = "";
}

my $threads = $opts{'t'};
my $rb_output     = "$output_prefix.RBs";
my $main_output   = "$output_prefix.tsv";
# Get the first contig/chr:
open(REFI,"<$ref_genome.fai") or die("Couldn't open reference index\n");
my $region        = ( split(/\t/,<REFI>) )[0];
close(REFI);

##########################################################################################
# Calculating number of reads and duplicate rates
my($num_unique_reads,$num_sequenced_reads,$dup_rate,$on_near_frac);
my($num_sequenced_reads_on_near, $num_unique_reads_on_near );

sub read_count {
    my ($threads, $bam, $panel, $extra_opts) = @_;
    $extra_opts = q{} unless(defined $extra_opts);
    my $tgt_opt = $panel eq "" ? "" : "-L $panel";
    $tgt_opt   .= " -f 2 -F 2828";
    print STDOUT "Calculating number of reads in $bam...\n";
    my $cmd = sprintf "samtools view $tgt_opt $extra_opts -@ %d -c %s", $threads, $bam;
    my ($stdout, $stderr, $exit) = capture {
        system($cmd);
    };
    die "Error calling $cmd, $stderr\n" if ( $exit != 0 );
    chomp( $stdout);
    return $stdout; # reads counted
}

$num_sequenced_reads = &read_count($threads, $merged_bam, "");
print STDOUT "  Num sequenced reads=$num_sequenced_reads\n";
if ( $do_panel ) {
  $num_sequenced_reads_on_near = &read_count($threads, $merged_bam, $panel);
  $num_unique_reads_on_near = &read_count($threads, $deduplicated_bam, $panel);
  $dup_rate = ($num_sequenced_reads_on_near-$num_unique_reads_on_near)/$num_sequenced_reads_on_near;
  $on_near_frac = $num_sequenced_reads_on_near/$num_sequenced_reads;
  print STDOUT "  Num sequenced reads (on+near)=$num_sequenced_reads_on_near\n";
  print STDOUT "  Num unique reads(on+near)=$num_unique_reads_on_near\n";
  print STDOUT "  Duplicate rate=$dup_rate\n";
  print STDOUT "  On+near fraction=$on_near_frac\n";
} else {
  $num_unique_reads = &read_count($threads, $deduplicated_bam, "");
  $dup_rate = ($num_sequenced_reads-$num_unique_reads)/$num_sequenced_reads;

  print STDOUT "  Num unique reads=$num_unique_reads\n";
  print STDOUT "  Duplicate rate=$dup_rate\n";
}


##########################################################################################
# Get read bundle comformations
my $bam = $merged_bam;
sub count_RB {
  my %rbs;
  my ($threads, $flag, $bam, $region, $panel) = @_;
  my $tgt_opt = $panel eq "" ? "" : "-L $panel";
  $tgt_opt   .= " -F 2828";
  my $cmd = "samtools view $tgt_opt -@ $threads -f $flag $bam $region";
  open(IN, "$cmd |") || die "Error launching $cmd\n"; 
  while(<IN>) {
    chomp;
    my @tmp = (split(/\t/,$_));
    my $rb;
    foreach my $t ( @tmp ) {
      if ( $t =~ /^RB:Z/ ) {
        $rb = $t;
        last;
      }
    }
    next unless(defined($rb));
    $rbs{$rb}++;
  }
  close(IN) or die ("error when calling $cmd : $?, $!\n");
  return( %rbs );
}
# Get first reads in reverse:
print STDOUT "RB comformation: 1st reads in reverse...\n";
my %rbf2r1 = &count_RB($threads, 82, $bam, $region, $panel);

print STDOUT "RB comformation: 2nd reads in reverse...\n";
my %rbf1r2 = &count_RB($threads, 146, $bam, $region, $panel);

foreach my $ikey (keys %rbf2r1) {
  $rbf1r2{$ikey} = 0 if ( ! exists( $rbf1r2{$ikey}));
}

foreach my $ikey (keys %rbf1r2) {
  $rbf2r1{$ikey} = 0 if ( ! exists( $rbf2r1{$ikey}));
}

open(OUT,">$rb_output") || die "Error openning output $rb_output\n";
foreach my $rb ( keys %rbf1r2 ) {
  print OUT "$rb\t",$rbf1r2{$rb},"\t",$rbf2r1{$rb},"\n";
}
close(OUT);

##########################################################################################
# Call R to get the two values that inform on strand misses:
my($reads_per_rb,$f_eff,$zib_eff,$ok_rbs,$total_rbs,$gc_both,$gc_single,$total_reads);
my $cmd = "efficiency_nanoseq.R $rb_output $ref_genome ";
print STDOUT "Running: $cmd\n";
my ($stdout, $stderr, $exit) = capture {
  system($cmd);
};
die "Error calling $cmd, $stderr\n" if ( $exit != 0 );
foreach (split('\n',$stdout)) {
  print STDOUT "  Routput: ",$_,"\n";
  chomp;
  if(/READS_PER_RB/) {
    $reads_per_rb = (split)[1];
  } elsif(/F-EFF/) {
    $f_eff = (split)[1];
  } elsif(/ZIB-EFF/) {
    $zib_eff = (split)[1];
  } elsif(/OK_RBS/) {
    $ok_rbs = (split)[1];
  } elsif(/TOTAL_RBS/) {
    $total_rbs = (split)[1];
  } elsif(/GC_BOTH/) {
    $gc_both = (split)[1];
  } elsif(/GC_SINGLE/) {
    $gc_single = (split)[1];
  } elsif(/TOTAL_READS/) {
    $total_reads = (split)[1];
  }
}
# Check it run properly:
open(OUT, ">$main_output") || die "Error writing to $main_output\n";
print OUT "# Whole-genome metrics:\n";
print OUT "NUM_READS_SEQUENCED\t$num_sequenced_reads\n";
if ($do_panel) {
  print OUT "ON+NEAR_FRACTION\t$on_near_frac\n";
  print OUT "ON+NEAR_UNIQUE_READS\t$num_unique_reads_on_near\n";
  print OUT "ON+NEAR_SEQUENCED_READS\t$num_sequenced_reads_on_near\n";
} else {
  print OUT "NUM_UNIQUE_READS\t$num_unique_reads\n";
}
print OUT "DUPLICATE_RATE\t$dup_rate\n";

print OUT "# RB metrics are reported for chr/contig $region only:\n";

my $bases_sequenced = $total_reads * 150;
my $bases_ok_rbs    = $ok_rbs * (300); # assuming mates don't overlap. Removing 50 bps for varios trimmings (rough estimate)
print OUT "TOTAL_RBS\t$total_rbs\n";
print OUT "TOTAL_READS_IN_RBS\t$total_reads\n";
print OUT "OK_RBS(2+2)\t$ok_rbs\n";
print OUT "READS_PER_RB\t$reads_per_rb\n";
print OUT "F-EFF\t$f_eff\n";
print OUT "EFFICIENCY\t",$bases_ok_rbs / $bases_sequenced,"\n";
print OUT "GC_BOTH\t$gc_both\n";
print OUT "GC_SINGLE\t$gc_single\n";
close(OUT);

__END__
=head1 NAME

efficiency_nanoseq.pl

=head1 SYNOPSIS

efficiency_nanoseq.pl [-h] [-t n] [-p panel ] -dedup BAM -duplex BAM -o prefix -r reference

    -dedup             -d   Deduplicated BAM
    -duplex            -x   Duplex BAM
    -ref               -r   reference file
    -out               -o   Ouptup prefix
    -panel             -p   Target panel (bed)

  Optional parameters:
    -threads           -t   Threads (1)
    -help              -h

=head1 OPTIONS

=over 8

=item B<-dedup>

Deduplicated BAM (neat) and index. The BAM after being processed by randomreadinbundle.

=item B<-duplex>

Duplex BAM and index. BAM prior to being processed by randomreadinbundle.

=item B<-out>

Output prefix

=item B<-ref>

Reference file and index

=item B<-panel>

Target panel (bed format)

=item B<-threads>

Use samtools with these many threads

=back

=cut

File: R/efficiency_nanoseq.R
#!/usr/bin/env Rscript

########## LICENCE ##########
# Copyright (c) 2022 Genome Research Ltd
# 
# Author: CASM/Cancer IT <cgphelp@sanger.ac.uk>
# 
# This file is part of NanoSeq.
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
# 
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
# 
# 1. The usage of a range of years within a copyright statement contained within
# this distribution should be interpreted as being equivalent to a list of years
# including the first and last year specified and all consecutive years between
# them. For example, a copyright statement that reads ‘Copyright (c) 2005, 2007-
# 2009, 2011-2012’ should be interpreted as being identical to a statement that
# reads ‘Copyright (c) 2005, 2007, 2008, 2009, 2011, 2012’ and a copyright
# statement that reads ‘Copyright (c) 2005-2012’ should be interpreted as being
# identical to a statement that reads ‘Copyright (c) 2005, 2006, 2007, 2008,
# 2009, 2010, 2011, 2012’.
###########################

suppressPackageStartupMessages({
  library(VGAM)
  library(Biostrings)
  library(Rsamtools)
  library(GenomicRanges)
  library(stringr)
  library(seqinr)
})

args = commandArgs(trailingOnly = TRUE)

if (length(args) == 0) {
  cat("efficiency_nanoseq.R rb_file genome\n\nMust specify a read-bundle file and a genome.\n\n")
  quit(save = "no", status = 0)
}

if (length(args) != 2) {
  cat("efficiency_nanoseq.R rb_file genome\n\nMust specify a read-bundle file and a genome.\n\n")
  quit(save = "no", status = 1)
}

rb_file = args[1]
genomeFile = args[2]

if (!file.exists(genomeFile)) {
  stop("Reference file not found : ", genomeFile, call. = FALSE)
}

if (!file.exists(rb_file)) {
  stop("readbundle file not found : ", rb_file, call. = FALSE)
}

pdf(paste(rb_file, ".pdf", sep = ""), width = 6, height = 6)
par(mar = c(2, 2, 2, 2))
cat("Reading file:", rb_file, "...\n")
rbs = read.table(rb_file, sep = "\t", header = F, row.names = 1)

rbs$x_tmp = pmin(rbs[, 1], 10)
rbs$y_tmp = pmin(rbs[, 2], 10)
rbs$x = rbs[,1]
rbs$y = rbs[,2]
rbs_bck = rbs;
kk = as.data.frame(table(rbs[, c("x", "y")]))
kk$point_size = kk$Freq / max(kk$Freq)
reads_per_RB = sum(c(rbs$x_tmp, rbs$y_tmp)) / nrow(rbs)
plot(as.numeric(kk$x),
       as.numeric(kk$y),
       pch = 20,
#cex=kk$point_size*10+.5,
       cex = kk$point_size * 10,
       xlab = "",
       ylab = "",
       main = paste(rb_file, ":", round(reads_per_RB, 3), sep = ""))
# Count how many do we have with size >= 4
# Define sizes 4, 5..., >=10
# For each size estimate how many we expect to have one strad ( 12.5% for 4, 6.25 for 5...)
rbs$size = rbs$x + rbs$y
rbs$size = pmin(rbs$size, 10)
total_missed = 0
for (size in c(4:10)) {
  exp_orphan = (0.5 ** size) * 2
  total_this_size = nrow(rbs[which(rbs$size == size),])
  if (total_this_size > 0) {
    with_both_strands = nrow(rbs[which(rbs$size == size & rbs$x > 0 & rbs$y > 0),])
    obs_orphan = 1 - with_both_strands / total_this_size
    missed = (obs_orphan - exp_orphan) * total_this_size
    total_missed = total_missed + missed
  }
}
total_missed_fraction = total_missed / nrow(rbs[which(rbs$size >= 4),])
rbs = rbs[which(rbs$size >= 2),]
rbs = rbs[sample(1:nrow(rbs), min(100000, nrow(rbs))),]
rbs$min = apply(rbs[, c("x", "y")], 1, min)
rbs$max = apply(rbs[, c("x", "y")], 1, max)
rbs$y = rbs$min
rbs$x = rbs$max
dev.off()

cat("READS_PER_RB\t", reads_per_RB, "\n", sep = "")
cat("F-EFF\t", total_missed_fraction, "\n", sep = "")
cat("OK_RBS\t", nrow(rbs_bck[which(rbs_bck$x >= 2 & rbs_bck$y >= 2),]), "\n", sep = "")
cat("TOTAL_RBS\t", nrow(rbs_bck), "\n", sep = "")
cat("TOTAL_READS\t", sum(c(rbs_bck$x, rbs_bck$y)) * 2, "\n", sep = "") #by 2 because we look only at one of the mates

##########################################################################################
# Now GC content:
#rbs = rbs_bck

colnames(rbs)[1:2] = c("plus", "minus")
rbs$id = rownames(rbs)
kk = str_split_fixed(rbs$id, "[:,]", 7)
rbs$chr = kk[, 3]
rbs$start = as.numeric(kk[, 4])
rbs$end = as.numeric(kk[, 5])

# Remove RBs with end breakpoints beyond chr/contig size
chr_coords = as.data.frame(scanFaIndex(genomeFile))
rownames(chr_coords) = chr_coords$seqnames
rbs$chr_end = chr_coords[rbs$chr, "end"]
rbs = rbs[which(rbs$end < rbs$chr_end),]

rbs_both = rbs[which(rbs$minus + rbs$plus >= 4 & rbs$minus >= 2 & rbs$plus >= 2),]
rbs_both = rbs_both[sample(1:nrow(rbs_both), min(10000, nrow(rbs_both))),] # 1000 random

rbs_single = rbs[which(rbs$minus + rbs$plus > 4 & (rbs$minus == 0 | rbs$plus == 0)),]
rbs_single = rbs_single[sample(1:nrow(rbs_single), min(10000, nrow(rbs_single))),] # 1000 random

seqs_both = as.vector(scanFa(genomeFile, GRanges(rbs_both$chr, IRanges(start = rbs_both$start, end = rbs_both$end))))
seqs_single = as.vector(scanFa(genomeFile, GRanges(rbs_single$chr, IRanges(start = rbs_single$start, end = rbs_single$end))))

seqs_both_collapsed = paste(seqs_both, collapse = "")
seqs_single_collapsed = paste(seqs_single, collapse = "")

tri_both = trinucleotideFrequency(DNAString(seqs_both_collapsed))
tri_single = trinucleotideFrequency(DNAString(seqs_single_collapsed))

tri_both_freqs = tri_both / sum(tri_both)
tri_single_freqs = tri_single / sum(tri_single)
if ( seqs_both_collapsed == "" ) {
  gc_both = 0
} else {
  gc_both = GC(s2c(seqs_both_collapsed))
}
if ( seqs_single_collapsed == "" ) {
  gc_single = 0
} else {
  gc_single = GC(s2c(seqs_single_collapsed))
}

cat("GC_BOTH\t", gc_both, "\n", sep = "")
cat("GC_SINGLE\t", gc_single, "\n", sep = "")

##########################################################################################
# RB sizes vs GC content & insert sizes:
rbs = rbs_bck


colnames(rbs)[1:2] = c("plus", "minus")
rbs$id = rownames(rbs)
rbs$size = rbs$plus + rbs$minus
kk = str_split_fixed(rbs$id, "[:,]", 7)
rbs$chr = kk[, 3]
rbs$start = as.numeric(kk[, 4])
rbs$end = as.numeric(kk[, 5])
rbs$insert = rbs$end - rbs$start

# Remove RBs with end breakpoints beyond chr/contig size
chr_coords = as.data.frame(scanFaIndex(genomeFile))
rownames(chr_coords) = chr_coords$seqnames
rbs$chr_end = chr_coords[rbs$chr, "end"]
rbs = rbs[which(rbs$end < rbs$chr_end),]


res = as.data.frame(matrix(nrow = 7, ncol = 3))
colnames(res) = c("size", "GC", "insert_len")
rownames(res) = c(1:7)
res[] = NA
for (size in c(1:7)) {
  rbs_ = rbs[which(rbs$size == size),]
  if (nrow(rbs_) == 0) {
    next
  }
  res[size, "size"] = size
  res[size, "insert_len"] = mean(rbs_$end - rbs_$start)
  rbs_ = rbs_[sample(1:nrow(rbs_), min(10000, nrow(rbs_))),] # 10000 random

  seqs_ = as.vector(scanFa(genomeFile, GRanges(rbs_$chr, IRanges(start = rbs_$start, end = rbs_$end))))

  seqs_collapsed = paste(seqs_, collapse = "")

  tri_ = trinucleotideFrequency(DNAString(seqs_collapsed))

  tri_freqs = tri_ / sum(tri_)

  gc_ = GC(s2c(seqs_collapsed))
  res[size, "GC"] = gc_

}
write.table(res, file = paste(rb_file, ".GC_inserts.tsv", sep = ""), row.names = F, col.names = T, sep = "\t", quote = F)


